spark2/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java

```diff
@@ -68,8 +68,8 @@ public DataSourceReader createReader(StructType readSchema, DataSourceOptions op
 
     Reader reader = new Reader(lazySparkSession(), table, Boolean.parseBoolean(caseSensitive), options);
     if (readSchema != null) {
-      // convert() will fail if readSchema contains fields not in table.schema()
-      SparkSchemaUtil.convert(table.schema(), readSchema);
+      // convert() will fail if readSchema contains fields not in reader.snapshotSchema()
+      SparkSchemaUtil.convert(reader.snapshotSchema(), readSchema);
```

Comment:
I think it would be better to expose `reader.snapshotId()` and use the snapshot ID to lookup the snapshot and its schema. This shouldn't be needed in Spark 3 so I think it is fine if it's a bit uglier here.

Reply:
Nevermind, I think it's okay the way it is working now.